build.ps1

```diff
@@ -33,36 +33,31 @@ if($MSBuildFullPath -ne "")
     }
 }
 
-# msbuild path not provided, find msbuild for VS2019
+# msbuild path not provided, find msbuild for VS2022
 if($MSBuildFullPath -eq "")
 {
-    if (Test-Path "env:\ProgramFiles(x86)")
+    $progFilesPaths = ${env:\ProgramFiles(x86)}, ${env:\ProgramFiles}
+    foreach ($progFilesPath in $progFilesPaths)
     {
-        $progFilesPath =  ${env:ProgramFiles(x86)}
-    }
-    elseif (Test-Path "env:\ProgramFiles")
-    {
-        $progFilesPath =  ${env:ProgramFiles}
-    }
-
-    $VS2019InstallPath = join-path $progFilesPath "Microsoft Visual Studio\2019"
-    $versions = 'Community', 'Professional', 'Enterprise'
+        $VS2022InstallPath = join-path $progFilesPath "Microsoft Visual Studio\2022"
+        $versions = 'Preview', 'Community', 'Professional', 'Enterprise'
 
-    foreach ($version in $versions)
-    {
-        $VS2019VersionPath = join-path $VS2019InstallPath $version
-        $MSBuildFullPath = join-path $VS2019VersionPath "MSBuild\Current\Bin\MSBuild.exe"
-
-        if (Test-Path $MSBuildFullPath)
+        foreach ($version in $versions)
         {
-            break
+            $VS2022VersionPath = join-path $VS2022InstallPath $version
+            $testPath = join-path $VS2022VersionPath "MSBuild\Current\Bin\MSBuild.exe"
+
+            if (Test-Path $testPath)
+            {
+                $MSBuildFullPath = $testPath
+            }
         }
     }
 }
 
-if (!(Test-Path $MSBuildFullPath))
+if (($MSBuildFullPath -eq "") -or !(Test-Path $MSBuildFullPath))
```

Comment:
Is the `-eq ""` check still needed?

Reply:
I think it is more robust this way. If for loop above doesn't give us path, `$MSBuildFullPath` will be empty and we will error out on `Test-Path` (which will surface as some NPE-like error) check rather than entering this if and throwing descriptive error as intended.

Reply:
This is not a blocker but feels like a violation of the "fewest elements" rule to me https://martinfowler.com/bliki/BeckDesignRules.html. We are not unit testing this script, but if we were, a concrete way to test whether the additional condition should remain would be creating a test that can fail independently of the other condition in the same line. If such a test cannot be created, it would be a strong indication that the additional condition shouldn't exist.

Reply:
If we want ot remove one of the terms in the expression, it would actually make more sense to remove the other condition (`Test-Path`). The point of the expression is to make sure that we've resolved a valid path to the MSBuild by this point.

`Test-Path $MSBuildFullPath` verifies that "[all elements of the path exist](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.management/test-path?view=powershell-7.4)". It **assumes that the path is not an empty string**. If the path is not resolved (empty string) it will not return false - it will throw an error.

`$MSBuildFullPath -eq ""` verifies that the path is resolved by the prior logic.

The only way that the `$MSBuildFullPath` can be non-empty is if it already explicitly passed `Test-Path` check (see few lines above). Therefore submitting it to `Test-Path` check once more doesn't add much value. The original author clearly intended to return an informative error in case path resolution failed ("Unable to find MSBuild installed on this machine..."), but with `Test-Path` this error can't ever surface. I know that because I experienced this situation, and was initially puzzled by it. With `$MSBuildFullPath -eq ""` we check exactly the situation we want to check and return the intended message.

Reply:
Ah, that makes sense because the loop logic above changed as well.